CppImport/src/macro/ExpansionManager.cpp

```diff
@@ -0,0 +1,213 @@
+	{
+		// extract everything in parentheses of the expansion signature using a regular expression
+		QRegularExpression regex ("\\((.*)\\)", QRegularExpression::DotMatchesEverythingOption);
+		auto argumentsString = lexicalHelper_->unexpandedSpelling(sr);
+		auto match = regex.match(argumentsString);
+		auto arguments = match.captured(1).split(",");
+
+		// by default initialize meta call arguments to be reference expressions with the raw spelling at this expansion
+		for (auto i = 0; i < clang_->argumentNames(entry->definition).size(); i++)
+		{
+			auto actualArg = args->getUnexpArgument((unsigned int)i);
+			entry->metaCall->arguments()->append(new OOModel::ReferenceExpression(arguments[i]));
+			entry->argumentLocs.append(actualArg->getLocation());
+		}
+	}
+
+	expansions_.append(entry);
+}
+
+QVector<MacroExpansion*> ExpansionManager::expansions() { return expansions_; }
+
+
+void ExpansionManager::clear()
+{
+	expansionCache_.clear();
+	expansions_.clear();
+}
+
+QVector<MacroExpansion*> ExpansionManager::topLevelExpansions()
+{
+	QVector<MacroExpansion*> result;
+	for (auto expansion : expansions_)
+		if (!expansion->parent)
+			result.append(expansion);
+
+	return result;
+}
+
+
+MacroExpansion*ExpansionManager::expansion(clang::SourceLocation loc)
+{
+	MacroExpansion* expansion = immediateExpansion(loc);
+	MacroExpansion* last = expansion;
+
+	if (expansion)
+	{
+		do
+		{
+			last = expansion;
+			loc = clang_->sourceManager()->getImmediateExpansionRange(loc).first;
+			expansion = immediateExpansion(loc);
+		} while (expansion && expansion->isChildOf(last));
+	}
+
+	return last;
+}
+
+MacroExpansion*ExpansionManager::immediateExpansion(clang::SourceLocation loc)
+{
+	auto expansion = clang_->immediateMacroLoc(loc);
+	for (auto i = 0; i < expansions_.size(); i++)
+		if (expansions_[i]->range.getBegin() == expansion) return expansions_[i];
+
+	/*
+	 * if we have not found an immediate expansion for loc then we try again using the found immediateExpansionLoc.
+	 * this can happen in case of token concatenation or stringifycation where the first expansion location would point
+	 * to the location of the concatenated token or stringifycation result.
+	 */
+	expansion = clang_->immediateMacroLoc(expansion);
+	for (auto i = 0; i < expansions_.size(); i++)
+		if (expansions_[i]->range.getBegin() == expansion) return expansions_[i];
+
+	return nullptr;
+}
+
+
+QSet<MacroExpansion*> ExpansionManager::expansion(Model::Node* node)
+{
+	Q_ASSERT(node);
+
+	if (!expansionCache_.contains(node))
+	{
+		expansionCache_[node] = {};
+
+		if (auto n = astMapping_->closestParentWithAstMapping(node))
+			if (astMapping_->contains(n))
+				for (auto range : astMapping_->get(n))
+				{
+					auto exp = expansion(range.getBegin());
+					if (exp)	expansionCache_[node].insert(exp);
+				}
+	}
+
+	return expansionCache_[node];
+}
+
+
+QVector<Model::Node*> ExpansionManager::tLExpansionTLNodes(MacroExpansion* expansion)
+{
+	Q_ASSERT(expansion);
```

Comment:
You should really try to move this code to `MacroExpansion`. It seems like you need to use some of the fields of `ExpansionManager`. Would it make sense, that each `MacroExpansion` knows its manager? Actually, is there ever more than one manager? If not, consider making it static.